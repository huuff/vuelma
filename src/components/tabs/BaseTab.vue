<template>
<slot></slot>
</template>

<script setup lang="ts">
import { FontAwesomeIconName } from '@/types/fontawesome-icon-name';

const props = defineProps<{
  titleText: string;
  tabId?: string;
  icon?: FontAwesomeIconName
}>();
</script>
