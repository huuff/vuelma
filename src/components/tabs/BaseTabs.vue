<template>
<render></render>
</template>

<script setup lang="tsx">
import { useSlots, VNode, toRef } from 'vue';
import { useOptionalTwoWayBinding } from '@/composables/optional-two-way-binding';
import { FontAwesomeIconName } from '@/types/fontawesome-icon-name';
import { iconAsRender } from '@/util/fontawesome-icon-render';
import BaseTab from './BaseTab.vue';
import partial from "lodash/partial";


const props = defineProps<{
  activeTabId?: string;
}>();

const emit = defineEmits<{
  (event: "update:activeTabId", tabId?: string): void;
}>();

const actualActiveTabId = useOptionalTwoWayBinding(undefined, toRef(props, "activeTabId"), partial(emit, "update:activeTabId"));

class Tab {
  public readonly tabId: string;

  constructor(
    public readonly title: string,
    public readonly icon: FontAwesomeIconName | undefined,
    public readonly slot: VNode<any, any, any>,
    tabId?: string,
  ) {
    this.tabId = tabId ?? title;
  }

  public getClasses(): { class: string } | Record<string, never> {
    if (this.tabId === actualActiveTabId.value) {
      return { class: "is-active"}; 
    } else {
      return {}
    }
  }

  public setActive(): void {
    actualActiveTabId.value = this.tabId;
  }

  public hasIcon(): boolean {
    return this.icon !== undefined;
  }
}

const slots = useSlots();

function slotToTabs(): Tab[] {
  const children = slots.default!();

  for (const child of children) {
    if (child.type !== BaseTab) {
      throw new Error(`All children of BaseTabs must be of type BaseTab, found ${child.type.toString()}`);
    }
  }

  return children.map(child => new Tab(
    child.props!.title,
    child.props!.icon,
    child,
    child.props!.tabId,
  ));
}

const render = () => {
  const tabs = slotToTabs();

  const activeTab = tabs.find(tab => tab.tabId === actualActiveTabId.value);

  return [
    <div class="tabs">
      <ul>
        { tabs.map(tab => 
          <li class={tab.getClasses}>
            <a onClick={() => tab.setActive()}>
              <span> {
                !tab.hasIcon()
                ? tab.title
                : [
                  iconAsRender(tab.icon!, [ "is-small" ]),
                  <span> { tab.title } </span>
                ]
              } </span>
            </a>
          </li>
        ) }
      </ul>
    </div>,
    activeTab?.slot
  ];
};
</script>
