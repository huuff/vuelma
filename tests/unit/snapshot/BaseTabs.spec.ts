import { mount } from "@vue/test-utils";
import BaseTabs from "@/components/tabs/BaseTabs.vue";
import BaseTab from "@/components/tabs/BaseTab.vue";

test("BaseTabs.vue snapshot", () => {
  const wrapper = mount(BaseTabs, {
    props: {
      activeTabId: "Tab 1",
    },
    global: {
      components: { BaseTab },
    },
    slots: {
      default: `
        <base-tab titleText="Tab 1">
          <p>Tab 1 contents</p>
        </base-tab>
        <base-tab titleText="Tab 2">
          <p>Tab 2 contents</p>
        </base-tab>
        <base-tab titleText="Tab 3">
          <p>Tab 3 contents</p>
        </base-tab>
      `
    },
  });

  expect(wrapper.element).toMatchSnapshot();
});
