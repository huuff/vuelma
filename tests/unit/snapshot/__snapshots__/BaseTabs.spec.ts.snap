// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTabs.vue snapshot 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="tabs"
  >
    <ul>
      
      <li
        class="is-active"
      >
        <a>
          <span>
            Tab 1
          </span>
        </a>
      </li>
      <li
        class=""
      >
        <a>
          <span>
            Tab 2
          </span>
        </a>
      </li>
      <li
        class=""
      >
        <a>
          <span>
            Tab 3
          </span>
        </a>
      </li>
      
    </ul>
  </div>
  
  <p>
    Tab 1 contents
  </p>
  
  
</div>
`;
